### PR TITLE
Remove dependency of psa crypto dependency on software AES

### LINKED ISSUE #3675: Break the dependency of psa_crypto on software AES
Currently `MBEDTLS_PSA_CRYPTO_C` depends on `MBEDTLS_CTR_DRBG_C` which depends on `MBEDTLS_AES_C`. The goal of this task is to remove or work around this dependency in order to allow using the PSA API with an AES accelerator.

Allowing the use of an alternate DRBG in PSA crypto is one way to do it, and is a desirable, but it is not a requirement of the present task. Note that [using HMAC_DRBG](https://github.com/ARMmbed/mbedtls/issues/3354) is not a solution for the present task: the goal here is to leverage an AES accelerator.

Some possible ways:

* Define a PSA DRBG abstraction and allow the PSA code to use it.
* Provide a way to plug an AES accelerator under `MBEDTLS_AES_ALT` that is at least sufficient for `ctr_drbg.c`.
* Allow `psa_generate_random` to call an RNG driver (not an entropy driver), which has the advantage of not going through the software CTR_DRBG, but the downside that it only benefits PSA interfaces (including symmetric key generation), not mbedtls implementations that use a DRBG (asymmetric key generation, blinding, non-deterministic algorithms).

### PR DESCRIPTION
**Pull request by jdurkop**
This set of changes is the framework for the removal of the psa_crypto dependency on software AES. The approach taken to perform this step was to create a wrapper in the psa crypto library for the mbedtls_ctr_drbg function calls so that a psa specific version of those calls can use AES accelerators once the driver supports them. It will also support backwards compatibility and still make the mbedtls_ctr_drbg calls if the MBEDTLS_PSA_ACCEL_ALG_AES_DRBG is not defined.

A new test has been added called build_psa_want_alg_dbrg_disabled_software that confirms the psa crypto library can still be built when PSA_WANT_ALG_AES_DRBG is defined with MBEDTLS_PSA_ACCEL_AES_DRBG and MBEDTLS_AES_C and MBEDTLS_CTR_DRBG_C are disabled. This essentially ensures that the dependency is removed if an alternative AES accelerator is available.

Fix #3675

## Status
READY

## Requires Backporting
NO

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
CI on 70d7f557b155734fb85f50b4acb7afd10ebbe8dd:

* Please run `scripts/generate_psa_constants.py`
* `test_ref_configs` fails on `config-symmetric-only.h`:
    ```
    /var/lib/build/library/psa_crypto.c:156:6: error: no previous prototype for 'psa_ctr_drbg_init' [-Werror=missing-prototypes]
     void psa_ctr_drbg_init( psa_ctr_drbg_context *ctx )
          ^
    /var/lib/build/library/psa_crypto.c:162:6: error: no previous prototype for 'psa_ctr_drbg_free' [-Werror=missing-prototypes]
     void psa_ctr_drbg_free( psa_ctr_drbg_context *ctx )
          ^
    /var/lib/build/library/psa_crypto.c:168:5: error: no previous prototype for 'psa_ctr_drbg_seed' [-Werror=missing-prototypes]
     int psa_ctr_drbg_seed( psa_ctr_drbg_context *ctx,
         ^
    /var/lib/build/library/psa_crypto.c:182:5: error: no previous prototype for 'psa_ctr_drbg_random' [-Werror=missing-prototypes]
     int psa_ctr_drbg_random( void *p_rng, unsigned char *output,
         ^
    ```

File: include/psa/crypto_values.h [hunk 1]
```diff
@@ -395,6 +395,7 @@
 /** Key for a cipher, AEAD or MAC algorithm based on the
  * Camellia block cipher. */
 #define PSA_KEY_TYPE_CAMELLIA                       ((psa_key_type_t)0x2403)
+#define PSA_KEY_TYPE_ARIA                           ((psa_key_type_t)0x2402)
```
Comment (gilles-peskine-arm) [inline include/psa/crypto_values.h:398]: Please don't declare ARIA in this pull request. It's unrelated. The new value isn't even acceptable: all key types must have even parity (this allows an implementation to defend against single-bit fault injection).

File: library/gcm.c [hunk 2]
```diff
@@ -1010,6 +1010,12 @@ int mbedtls_gcm_self_test( int verbose )
     return( ret );
 }
 
+#else
+int mbedtls_gcm_self_test( int verbose )
```
Comment (gilles-peskine-arm) [inline library/gcm.c:1014]: No, the GCM self-test function should not be defined if it doesn't do anything.

File: library/psa_crypto.c [hunk 3]
```diff
@@ -93,7 +93,102 @@ static inline int safer_memcmp( const uint8_t *a, const uint8_t *b, size_t n )
...
+
+#if defined(MBEDTLS_THREADING_C)
+    mbedtls_threading_mutex_t mutex;
+#endif
+}
+psa_ctr_drbg_context;
+
+/*
+ * TBD: These functions need to be filled out with the support required to
+ * use driver acceleration for DRBG support.
```
Comment (gilles-peskine-arm) [inline library/psa_crypto.c:154]: Yes, they do. I'm confused: is the PR ready or not? I don't see how the PR is useful without implementing these functions.

File: library/psa_crypto.c [hunk 4]
```diff
@@ -93,7 +93,102 @@ static inline int safer_memcmp( const uint8_t *a, const uint8_t *b, size_t n )
...
+                                 * This is the number of requests that have
+                                 * been made since the last (re)seeding,
+                                 * minus one.
+                                 * Before the initial seeding, this field
+                                 * contains the amount of entropy in bytes
+                                 * to use as a nonce for the initial seeding,
+                                 * or -1 if no nonce length has been explicitly
+                                 * set (see mbedtls_ctr_drbg_set_nonce_len()).
+                                 */
+    int prediction_resistance;  /*!< This determines whether prediction
```
Comment (gilles-peskine-arm) [inline library/psa_crypto.c:126]: Why include all the fields? This module only uses ctr_drbg in a specific configuration. `prediction_resistance` isn't used. `mbedtls_aes_context` should be an accelerator-specific type.

File: tests/suites/test_suite_cipher.aes.data [hunk 5]
```diff
@@ -4,7 +4,7 @@ dec_empty_buf:MBEDTLS_CIPHER_AES_128_CBC:0:0
 
 AES-128 XTS - Decrypt empty buffer
 depends_on:MBEDTLS_AES_C:MBEDTLS_CIPHER_MODE_XTS
-dec_empty_buf:MBEDTLS_CIPHER_AES_128_XTS:MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH:0
+dec_empty_buf:MBEDTLS_CIPHER_AES_128_XTS:-0x22:0
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_cipher.aes.data:7]: Where does this numerical value come from? Please use symbolic constant names. This test case depends on `MBEDTLS_AES_C` so `MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH` is guaranteed to be defined.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I'm confused about the status of this PR. It's incomplete and isn't mergeable in its current state.

**Comment by jdurkop** [comment]
@gilles-peskine-arm Would you like me to pull this PR back and do additional work? I thought the point of this work was similar to the other tickets in this epic. Simply laying the ground work to remove builtin libraries when accelerator options are requested. I assumed that the actual accelerator support would be added in with a different ticket. If that is not the case then you are correct and this is not ready. Some of the changes outside of psa_crypto.c were done to correct CI build failures that were occurring when MBEDTLS_AES_C is not defined for the new test added. I will look into different approaches on resolving those failures.

**Comment by gilles-peskine-arm** [comment]
The goal is to allow the DRBG used by `psa_crypto.c` to leverage PSA AES acceleration, either by defining a PSA DRBG interface or by using the existing interface for `psa_cipher_xxx` with AES-ECB. This PR doesn't have either a PSA DRBG interface (with code in `psa_crypto_driver_wrappers.c`) or the code to call `psa_cipher_xxx`.

I don't mind which approach is chosen (both are valuable, and since Silabs is doing the work, you get to pick which one gets implemented first). But the result has to be usable.

**Comment by jdurkop** [comment]
@gilles-peskine-arm Ok. Sounds like the scope is larger than what I was thinking. I'll close this out and open a new one when additional work is completed.

**Pull request closed without merging**